src/lib/utils/lib.ts

```diff
@@ -52,3 +56,132 @@ export const getTransactionReceipt = async (
 
 export const isDefined = <T>(val: T | null | undefined): val is T =>
   typeof val !== 'undefined' && val !== null
+
+export const isArbitrumChain = async (provider: Provider): Promise<boolean> => {
+  try {
+    await ArbSys__factory.connect(ARB_SYS_ADDRESS, provider).arbOSVersion()
+  } catch (error) {
+    return false
+  }
+  return true
+}
+
+/**
+ * This function performs a binary search to find the first L2 block that corresponds to a given L1 block number.
+ * The function returns a Promise that resolves to a number if a block is found, or undefined otherwise.
+ *
+ * @param {JsonRpcProvider} provider - The L2 provider to use for the search.
+ * @param {number} forL1Block - The L1 block number to search for.
+ * @param {boolean} [allowGreater=false] - Whether to allow the search to go past the specified `forL1Block`.
+ * @param {number|string} minL2Block - The minimum L2 block number to start the search from. Cannot be below the network's `nitroGenesisBlock`.
+ * @param {number|string} [maxL2Block='latest'] - The maximum L2 block number to end the search at. Can be a `number` or `'latest'`. `'latest'` is the current block.
+ * @returns {Promise<number | undefined>} - A Promise that resolves to a number if a block is found, or undefined otherwise.
+ */
+export async function getFirstBlockForL1Block({
+  provider,
+  forL1Block,
+  allowGreater = false,
+  minL2Block,
+  maxL2Block = 'latest',
+}: {
+  provider: JsonRpcProvider
+  forL1Block: number
+  allowGreater?: boolean
+  minL2Block?: number
+  maxL2Block?: number | 'latest'
+}): Promise<number | undefined> {
+  if (!isArbitrumChain(provider)) {
+    throw new Error('Arbitrum provider is required.')
```

Comment:
I suggest we assume the provider is L1 if its not Arbitrum, and return `forL1Block`

Reply:
can do